File: src/libnymble++-xpcom/nsNymbleUser.js
google.load('jquery', '1');

function getUTCDayOfYear(date) {   
  // d is a Date object
  var yn = date.getUTCFullYear();
  var mn = date.getUTCMonth();
  var dn = date.getUTCDate();
  var d1 = new Date(yn, 0, 1, 12, 0, 0); // noon on Jan. 1
  var d2 = new Date(yn, mn, dn, 12, 0, 0); // noon on input date
  var ddiff = Math.round((d2 - d1) / 864e5);
  return ddiff + 1; 
}

google.setOnLoadCallback(function() {
  netscape.security.PrivilegeManager.enablePrivilege('UniversalXPConnect');
  
  jQuery('form').each(function() {
    var form = this;
    var pseudonym_manager = jQuery('link[rel="nymble.pseudonym_manager"]').get(0).href;
    var nymble_manager    = jQuery('link[rel="nymble.manager"]').get(0).href;
    var server_blacklist  = jQuery('link[rel="nymble.blacklist"]').get(0).href;
    
    if (!pseudonym_manager || !nymble_manager || !server_blacklist || !confirm("Do you want to acquire a nymble to authenticate with?")) {
      return null;
    }
    
    var user = Components.classes['@nymble.cs.dartmouth.edu/nymble/NymbleUser;1'].createInstance().QueryInterface(Components.interfaces.nsINymbleUser);
    var cur_time = new Date();
    
    user.setLinkWindow(366 * (cur_time.getUTCFullYear() - 1970) + getUTCDayOfYear(cur_time));
    user.setTimePeriod(cur_time.getUTCHours() * 60 + cur_time.getUTCMinutes());
    
    jQuery('p#status').append('Getting pseudonym...');
    
    jQuery.post(pseudonym_manager, null, function(data, textStatus) {
      netscape.security.PrivilegeManager.enablePrivilege('UniversalXPConnect');
      
      user.setPseudonym(data.pseudonym);
      jQuery('p#status').append('done!<br/>Getting blacklist...');
      
      jQuery.get(server_blacklist, null, function(data, textStatus) {
        netscape.security.PrivilegeManager.enablePrivilege('UniversalXPConnect');
        
        server_id = user.addBlacklist(data.blacklist);
        jQuery('p#status').append('done!</br>Getting credential...');
        
        jQuery.get(nymble_manager + server_id + '/', { 'pseudonym': user.getPseudonym() }, function(data, textStatus) {
          netscape.security.PrivilegeManager.enablePrivilege('UniversalXPConnect');
          
          user.addCredential(data.credential);
          
          jQuery('p#status').append('done!');
          jQuery('input[id="nymble_ticket"]')[0].value = user.getTicket(server_id);
        }, 'json');
      }, 'json');
    }, 'json');
  });
});

File: src/libnymble++-xpcom/nsNymbleUser.cc
#include "nsNymbleUser.h"

NS_IMPL_ISUPPORTS1(nsNymbleUser, nsINymbleUser)

nsNymbleUser::nsNymbleUser()
{
  this->user = new Nymble::User();
}

nsNymbleUser::~nsNymbleUser()
{
  delete this->user;
}


/* void setTimePeriod (in PRUint32 time_period); */
NS_IMETHODIMP nsNymbleUser::SetTimePeriod(PRUint32 time_period)
{
  this->user->setTimePeriod(time_period);
  
  return NS_OK;
}

/* void setLinkWindow (in PRUint32 link_window); */
NS_IMETHODIMP nsNymbleUser::SetLinkWindow(PRUint32 link_window)
{
  this->user->setLinkWindow(link_window);
  
  return NS_OK;
}

/* void readVerifyKey (in string verify_key_path); */
NS_IMETHODIMP nsNymbleUser::ReadVerifyKey(const char *verify_key_path)
{
  this->user->readVerifyKey((char*) verify_key_path);
  return NS_OK;
}

/* void setPseudonym (in string marshalled_pseudonym); */
NS_IMETHODIMP nsNymbleUser::SetPseudonym(const char *marshalled_pseudonym)
{
  Nymble::Pseudonym* pseudonym = new Nymble::Pseudonym();
  
  Nymble::Pseudonym::unmarshal((char*) marshalled_pseudonym, pseudonym);
  
  this->user->setPseudonym(pseudonym);
  
  delete pseudonym;
  
  return NS_OK;
}

/* string getPseudonym (); */
NS_IMETHODIMP nsNymbleUser::GetPseudonym(char **_retval)
{
  Nymble::Pseudonym* pseudonym = this->user->getPseudonym();
  
  if (pseudonym == NULL) {
    *_retval = NULL;
  } else {
    *_retval = (char*) malloc(pseudonym->marshal() + 1);
    pseudonym->marshal(*_retval);
  }
  
  return NS_OK;
}

/* string addBlacklist (in string marshalled_blacklist); */
NS_IMETHODIMP nsNymbleUser::AddBlacklist(const char *marshalled_blacklist, char **_retval)
{
  Nymble::Blacklist* blacklist = new Nymble::Blacklist();
  
  Nymble::Blacklist::unmarshal((char*) marshalled_blacklist, blacklist);
  
  u_char* server_id = this->user->addBlacklist(blacklist);
  
  if (server_id == NULL) {
    *_retval = NULL;
  } else {
    *_retval = (char*) malloc(Nymble::hexencode(server_id, DIGEST_SIZE) + 1);
    Nymble::hexencode(server_id, DIGEST_SIZE, *_retval);
  }
  
  delete blacklist;
  
  return NS_OK;
}

/* boolean addCredential (in string marshalled_credential); */
NS_IMETHODIMP nsNymbleUser::AddCredential(const char *marshalled_credential, PRBool *_retval)
{
  Nymble::Credential* credential = new Nymble::Credential();
  
  Nymble::Credential::unmarshal((char*) marshalled_credential, credential);
  
  *_retval = this->user->addCredential(credential);
  
  delete credential;
  
  return NS_OK;
}


/* string getTicket (in string serverId); */
NS_IMETHODIMP nsNymbleUser::GetTicket(const char *serverId, char **_retval)
{
  u_int server_id_len = Nymble::hexdecode((char*) serverId);
  u_char server_id[server_id_len];
  
  Nymble::hexdecode((char*) serverId, server_id);

  Nymble::Ticket* ticket;
  
  if (server_id_len == DIGEST_SIZE) {
    ticket = this->user->getTicket(server_id);
  }
  
  if (ticket == NULL) {
    *_retval = NULL;
  } else {
    *_retval = (char*) malloc(ticket->marshal() + 1);
    ticket->marshal(*_retval);
  }
  
  return NS_OK;
}
